cdc_rsync/server_arch.cc

```diff
@@ -0,0 +1,157 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cdc_rsync/server_arch.h"
+
+#include <filesystem>
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_format.h"
+#include "common/log.h"
+#include "common/path.h"
+#include "common/remote_util.h"
+
+namespace cdc_ft {
+
+constexpr char kErrorFailedToGetKnownFolderPath[] =
+    "error_failed_to_get_known_folder_path";
+constexpr char kErrorArchTypeUnhandled[] = "arch_type_unhandled";
+
+// static
+ServerArch::Type ServerArch::Detect(const std::string& destination) {
+  // Path starting with ~ or / -> Linux.
+  if (absl::StartsWith(destination, "~") ||
+      absl::StartsWith(destination, "/")) {
+    return Type::kLinux;
+  }
+
+  // Path starting with C: etc. -> Windows.
+  if (!path::GetDrivePrefix(destination).empty()) {
+    return Type::kWindows;
+  }
+
+  // Path with only forward slashes -> Linux.
+  if (absl::StrContains(destination, "/") &&
+      !absl::StrContains(destination, "\\")) {
+    return Type::kLinux;
+  }
+
+  // Path with only forward backslashes -> Windows.
+  if (absl::StrContains(destination, "\\") &&
+      !absl::StrContains(destination, "/")) {
+    return Type::kWindows;
+  }
+
+  // Default to Linux.
+  return Type::kLinux;
+}
+
+ServerArch::ServerArch(Type type) : type_(type) {
+  absl::Status status =
+      path::GetKnownFolderPath(path::FolderId::kRoamingAppData, &appdata_path_);
+  if (!status.ok()) {
+    // Should be extremely rare.
+    LOG_ERROR("Failed to get roaming appdata path: %s", status.ToString());
+    appdata_path_ = kErrorFailedToGetKnownFolderPath;
+  }
+  path::EnsureEndsWithPathSeparator(&appdata_path_);
+}
+
+ServerArch::~ServerArch() {}
+
+std::string ServerArch::CdcServerFilename() const {
+  switch (type_) {
+    case Type::kWindows:
+      return "cdc_rsync_server.exe";
+    case Type::kLinux:
+      return "cdc_rsync_server";
+    default:
+      assert(!kErrorArchTypeUnhandled);
+      return std::string();
+  }
+}
+
+std::string ServerArch::RemoteToolsBinDir() const {
+  switch (type_) {
+    case Type::kWindows:
+      return appdata_path_ + "cdc-file-transfer\\bin\\";
```

Comment:
Oh good point. Yeah, that'll go wrong if the user names are different. I struggled a bit to fix it properly. While for ssh you can use $env:appdata (because ssh runs in Powershell, at least on our systems), for scp it is not possible to use %appdata% or $env:appdata. It seems like scp just doesn't expand shell variables. Note that doing something like `scp src user@host:%appdata%\dst` works in a cmd because it uses shell expansion in that cmd, but that's the wrong one.

I use "AppData\Roaming" relative to %UserProfile% (the default target location for scp) now. I'm not sure whether that always works, though. I'll look into sftp to see if we can replace scp entirely (would be nice).